File: server/presence.ts
/**
 * Who is looking at a track right now, and who is mid-sentence.
 *
 * Two facts with the same shape and very different lifetimes, which is the
 * whole design here:
 *
 *   **Watching** lasts 45 seconds and is refreshed by a heartbeat. It has to
 *   outlive a slow tab, a laptop lid, a garbage-collected timer — so it is
 *   generous, and the cost of being generous is that somebody who closed their
 *   laptop lingers for up to three quarters of a minute. That is the right
 *   trade: a name that hangs about briefly is a much smaller lie than a name
 *   that vanishes while its owner is still reading.
 *
 *   **Typing** lasts 3 seconds and is refreshed by keystrokes. It has to be
 *   *ungenerous* for exactly the opposite reason: "Ana is typing…" left up
 *   after Ana wandered off is worse than no indicator at all, because it is
 *   the one signal people wait on before sending something themselves.
 *
 * In process and in one map, which is fine on the single replica a SQLite
 * volume already forces. Nothing here is persisted, and nothing should be:
 * presence that survived a restart would be a list of people who are not
 * there.
 */
import type { Presence } from "../shared/api";
import { publish } from "./hub";

const WATCH_TTL_MS = 45_000;
const TYPING_TTL_MS = 3_000;
/** How often lapses are noticed. Half the typing window, so it never overshoots by much. */
const SWEEP_MS = 1_500;

interface Watcher {
  userId: string;
  login: string;
  name: string | null;
  avatarUrl: string | null;
  projectId: string;
  watchingUntil: number;
  typingUntil: number;
}

/**
 * One track's room: who is in it, and who may be told about it.
 *
 * The audience is held here rather than re-derived, because the sweeper — the
 * only thing that notices somebody lapsing — runs on a timer with no request
 * behind it, and a database read per track per tick to answer "who may know"
 * is a poor trade for a frame nobody asked for. It is refreshed on every beat,
 * so an invitation takes effect on the inviter's next heartbeat at the latest.
 */
interface Room {
  watchers: Map<string, Watcher>;
  projectId: string;
  audience: ReadonlySet<string>;
}

/** trackId → room. */
const tracks = new Map<string, Room>();

/**
 * A heartbeat from one browser.
 *
 * `typing` is a pulse rather than a state: the browser says "still typing" and
 * the server gives that three seconds, instead of the browser having to
 * remember to say "stopped". A browser that crashes mid-word therefore stops
 * claiming to be typing without having to apologise for it, which a
 * start/stop protocol could not manage.
 */
export function beat(input: {
  trackId: string;
  projectId: string;
  userId: string;
  login: string;
  name: string | null;
  avatarUrl: string | null;
  typing: boolean;
  audience: ReadonlySet<string>;
}): Presence[] {
  const now = Date.now();
  const room: Room = tracks.get(input.trackId) ?? { watchers: new Map(), projectId: input.projectId, audience: input.audience };
  room.projectId = input.projectId;
  room.audience = input.audience;
  const existing = room.watchers.get(input.userId);
  room.watchers.set(input.userId, {
    userId: input.userId,
    login: input.login,
    name: input.name,
    avatarUrl: input.avatarUrl,
    projectId: input.projectId,
    watchingUntil: now + WATCH_TTL_MS,
    // A heartbeat that is not a typing pulse must not cancel one that is:
    // the composer pings on a timer and on keystrokes independently, and the
    // slower of the two arriving second would blink the indicator off.
    typingUntil: input.typing ? now + TYPING_TTL_MS : (existing?.typingUntil ?? 0),
  });
  tracks.set(input.trackId, room);

  const here = present(input.trackId, now);
  // Published every beat rather than only on change: a beat is at most one
  // per browser per twenty seconds, and comparing sets to save a frame that
  // small is more code than it saves.
  publish(input.projectId, { event: "here", data: { trackId: input.trackId, present: here } }, input.audience);
  ensureSweeper();
  return here;
}

/** Who is on a track now, watchers first typing or not, self included. */
export function present(trackId: string, now = Date.now()): Presence[] {
  const room = tracks.get(trackId);
  if (!room) return [];
  const out: Presence[] = [];
  for (const w of room.watchers.values()) {
    if (w.watchingUntil <= now) continue;
    out.push({ login: w.login, name: w.name, avatarUrl: w.avatarUrl, typing: w.typingUntil > now });
  }
  return out.sort((a, b) => a.login.localeCompare(b.login));
}

/**
 * Somebody closing a track, rather than lapsing out of it.
 *
 * Worth having as well as the lease: leaving a track is the one moment we
 * genuinely know, and forty-five seconds of a ghost in the corner of somebody
 * else's screen is exactly the kind of small wrongness that makes a presence
 * feature feel broken.
 */
export function leave(trackId: string, userId: string, projectId: string, audience: ReadonlySet<string>): void {
  const room = tracks.get(trackId);
  if (!room?.watchers.delete(userId)) return;
  if (room.watchers.size === 0) tracks.delete(trackId);
  publish(projectId, { event: "here", data: { trackId, present: present(trackId) } }, audience);
}

/**
 * The one timer, started on first use and stopped when nobody is anywhere.
 *
 * A lapse is the only presence change that arrives with no request attached,
 * so something has to go looking for it. It publishes only when a room's
 * membership or typing state actually changed, because the alternative is a
 * frame every second and a half to every open browser forever.
 */
let sweeper: ReturnType<typeof setInterval> | null = null;
const lastPublished = new Map<string, string>();

function ensureSweeper(): void {
  if (sweeper) return;
  sweeper = setInterval(() => {
    const now = Date.now();
    if (tracks.size === 0) {
      clearInterval(sweeper!);
      sweeper = null;
      lastPublished.clear();
      return;
    }
    for (const [trackId, room] of [...tracks]) {
      for (const [userId, w] of [...room.watchers]) if (w.watchingUntil <= now) room.watchers.delete(userId);
      const { projectId, audience } = room;
      if (room.watchers.size === 0) tracks.delete(trackId);

      const here = present(trackId, now);
      const fingerprint = here.map((p) => `${p.login}${p.typing ? "!" : ""}`).join(",");
      if (lastPublished.get(trackId) === fingerprint) continue;
      if (fingerprint) lastPublished.set(trackId, fingerprint);
      else lastPublished.delete(trackId);
      publish(projectId, { event: "here", data: { trackId, present: here } }, audience);
    }
  }, SWEEP_MS);
  // Never hold the process open for a presence timer.
  sweeper.unref?.();
}

/** For tests: forget everybody. */
export function resetPresence(): void {
  tracks.clear();
  lastPublished.clear();
  if (sweeper) {
    clearInterval(sweeper);
    sweeper = null;
  }
}

export const PRESENCE_TIMINGS = { WATCH_TTL_MS, TYPING_TTL_MS, SWEEP_MS };

File: lib/ravix/presence.ex
defmodule Ravix.Presence do
  @moduledoc """
  Who is looking at a track right now, and who is mid-sentence.

  Two facts with the same shape and very different lifetimes, which is the
  whole design here:

    **Watching** lasted 45 seconds over HTTP and was refreshed by a heartbeat,
    because a stateless request had no other way to know a tab was still
    open. A LiveView is a process, so watching is now the process itself: a
    page tracks itself on the track's topic when it opens the track, and the
    tracker notices the moment it goes away, whether by navigation, a closed
    tab or a dropped socket. Nothing lingers for three quarters of a minute
    and nothing needs a timer to notice a lapse.

    **Typing** lasts 3 seconds and is refreshed by keystrokes. It has to be
    *ungenerous*: "Ana is typing..." left up after Ana wandered off is worse
    than no indicator at all, because it is the one signal people wait on
    before sending something themselves. A pulse rather than a state, so a
    browser that stops mid-word stops claiming to be typing without having to
    apologise for it, which a start/stop protocol could not manage.

  `Phoenix.Presence` over `Ravix.PubSub`, one topic per track. Every change
  to a track's room (a join, a leave, a typing pulse) reaches `handle_metas/4`
  in the presence server, which publishes the whole set as a `here` event on
  the project's hub topic (`shared/api.ts` `ProjectEvent`), and schedules one
  more frame for when the newest typing pulse lapses. Published every time
  rather than only on change: a frame is small, and comparing sets to save
  one is more code than it saves.

  Nothing here is persisted, and nothing should be: presence that survived a
  restart would be a list of people who are not there.

  The TypeScript scoped each `here` frame to an audience (the owner, the
  project's members and the track's own members) so an event naming a track
  did not tell a stranger the track exists. `Ravix.Hub` fans out by project,
  so the subscriber does that filtering: a page keeps a `here` frame only for
  a track it is already allowed to list. The frame carries logins, names and
  avatars and nothing else.
  """

  use Phoenix.Presence, otp_app: :ravix, pubsub_server: Ravix.PubSub

  alias Ravix.Accounts.User
  alias Ravix.Hub
  alias Ravix.Presence.Watcher

  @typing_ttl_ms 3_000

  @typedoc "Somebody with a track open right now; see `Ravix.Presence.Watcher`."
  @type presence :: Watcher.t()

  @typedoc """
  What a heartbeat reports. `:watching` is the composer's timer saying the
  page is still open; `:typing` is a keystroke. The two arrive
  independently, which is why they are named rather than flagged: a
  `beat(..., false)` read as cancelling a typing pulse, and it does not.
  """
  @type activity :: :typing | :watching

  @doc "How long a typing pulse stands."
  @spec typing_ttl_ms() :: pos_integer()
  def typing_ttl_ms, do: @typing_ttl_ms

  @doc "The presence topic for a track."
  @spec topic(String.t()) :: String.t()
  def topic(track_id), do: "presence:track:" <> track_id

  @doc """
  A heartbeat from the calling process: I am here, and possibly typing.

  The first beat from a process tracks it in the track's room; later beats
  update its metadata. A heartbeat that is not a typing pulse must not cancel
  one that is: the composer pings on a timer and on keystrokes independently,
  and the slower of the two arriving second would blink the indicator off.
  Returns who is in the room now, the caller included.
  """
  @spec beat(String.t(), String.t(), User.t(), activity()) :: [presence()]
  def beat(track_id, project_id, %User{} = user, activity)
      when activity in [:typing, :watching] do
    now = now_ms()
    topic = topic(track_id)

    meta = fn existing ->
      %{
        project_id: project_id,
        login: user.login,
        name: user.name,
        avatar_url: user.avatar_url,
        typing_until:
          if(activity == :typing,
            do: now + @typing_ttl_ms,
            else: Map.get(existing, :typing_until, 0)
          )
      }
    end

    case update(self(), topic, user.id, meta) do
      {:ok, _ref} -> :ok
      {:error, :nopresence} -> {:ok, _ref} = track(self(), topic, user.id, meta.(%{}))
    end

    present(track_id, now)
  end

  @doc """
  Somebody closing a track, rather than lapsing out of it.

  Worth having as well as the process lifetime: switching tracks in one page
  is the one moment we genuinely know, and a name in the corner of a
  colleague's screen that belongs to somebody now reading another track is
  exactly the kind of small wrongness that makes presence feel broken.
  """
  @spec leave(String.t(), String.t()) :: :ok
  def leave(track_id, user_id), do: untrack(self(), topic(track_id), user_id)

  @doc "Who is on a track now, sorted by login so the row never reshuffles, self included."
  @spec present(String.t(), integer()) :: [presence()]
  def present(track_id, now \\ now_ms()) do
    track_id |> topic() |> list() |> presences_of(now)
  end

  # ── the presence server's side ────────────────────────────────────────

  @impl true
  def init(_opts), do: {:ok, %{}}

  @impl true
  def handle_metas("presence:track:" <> track_id, diff, presences, state) do
    presences = Map.new(presences, fn {id, metas} -> {id, %{metas: metas}} end)
    now = now_ms()
    here = presences_of(presences, now)

    # The room may have just emptied, in which case the only meta that still
    # knows the project is the one that left. An empty set is still news.
    case project_id_of(presences) || project_id_of(Map.get(diff, :leaves, %{})) do
      nil ->
        {:ok, Map.delete(state, track_id)}

      project_id ->
        publish(project_id, track_id, here)

        if presences == %{},
          do: {:ok, Map.delete(state, track_id)},
          else: {:ok, schedule_lapse(state, track_id, project_id, presences, now)}
    end
  end

  def handle_metas(_topic, _diff, _presences, state), do: {:ok, state}

  # One delayed frame per typing window, so "typing" goes out when the pulse
  # lapses even though no request arrives to say so. A pulse that extends the
  # window past what is scheduled schedules again; the earlier frame still
  # goes out, and says the truth as of that moment.
  defp schedule_lapse(state, track_id, project_id, presences, now) do
    latest =
      presences
      |> Enum.flat_map(fn {_key, %{metas: metas}} ->
        Enum.map(metas, &Map.get(&1, :typing_until, 0))
      end)
      |> Enum.max(fn -> 0 end)

    if latest > now and latest > Map.get(state, track_id, 0) do
      delay = latest - now + 1

      Task.Supervisor.start_child(Ravix.TaskSupervisor, fn ->
        Process.sleep(delay)
        publish(project_id, track_id, present(track_id))
      end)

      Map.put(state, track_id, latest)
    else
      state
    end
  end

  # Local, not cluster-wide (#19). Every instance runs `handle_metas/4` on the
  # same diff, so a cluster-wide broadcast from here would reach each reader once
  # per instance -- measured at three frames for one change on two nodes. Each
  # instance serving its own readers gets everyone exactly one, and the frame is
  # the whole room either way, because `list/1` is cluster-wide.
  defp publish(project_id, track_id, here) do
    Hub.publish_local(project_id, :here, track_id: track_id, present: here)
  end

  defp project_id_of(presences) do
    Enum.find_value(presences, fn {_key, %{metas: metas}} ->
      Enum.find_value(metas, &Map.get(&1, :project_id))
    end)
  end

  # Several metas per person (two tabs) are one entry: typing if any of them is.
  defp presences_of(presences, now) do
    presences
    |> Enum.map(fn {_key, %{metas: [first | _] = metas}} ->
      %Watcher{
        login: first.login,
        name: first.name,
        avatar_url: first.avatar_url,
        typing: Enum.any?(metas, &(Map.get(&1, :typing_until, 0) > now))
      }
    end)
    |> Enum.sort_by(& &1.login)
  end

  defp now_ms, do: System.system_time(:millisecond)
end
